### DISCUSSION TITLE
Cannot schedule a task: cannot allocate thread

### DISCUSSION THREAD

**Discussion by shenjingmushan** [ClickHouse/ClickHouse#46097] [Q&A]
hi.
I encountered a error, when query ch, error stack: 

2023.02.07 13:46:08.665585 [ 30088 ] {7fd8cc18-01e1-4ee0-8a56-a4b4b6953372} <Error> executeQuery: Code: 439. DB::Exception: Cannot schedule a task: cannot allocate thread (threads=427, jobs=427). (CANNOT_SCHEDULE_TASK) (version 22.8.3.13 (official build)) (from 11.154.26.154:55695) (in query: select * from xxxx t final order by t.a desc limit 100), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long, unsigned long&>(int, fmt::v8::basic_format_string<char, fmt::v8::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>::type, fmt::v8::type_identity<unsigned long>::type, fmt::v8::type_identity<unsigned long&>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long&&, unsigned long&) in /usr/bin/clickhouse
2. void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const in /usr/bin/clickhouse
3. void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>) in /usr/bin/clickhouse
4. ? in /usr/bin/clickhouse
5. DB::PipelineExecutor::spawnThreads() in /usr/bin/clickhouse
6. DB::PipelineExecutor::executeImpl(unsigned long) in /usr/bin/clickhouse
7. DB::PipelineExecutor::execute(unsigned long) in /usr/bin/clickhouse
8. DB::CompletedPipelineExecutor::execute() in /usr/bin/clickhouse
9. DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::__1::shared_ptr<DB::Context>, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>, std::__1::optional<DB::FormatSettings> const&) in /usr/bin/clickhouse
10. DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::__1::optional<DB::CurrentThread::QueryScope>&) in /usr/bin/clickhouse
11. DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&) in /usr/bin/clickhouse
12. DB::HTTPServerConnection::run() in /usr/bin/clickhouse
13. Poco::Net::TCPServerConnection::start() in /usr/bin/clickhouse
14. Poco::Net::TCPServerDispatcher::run() in /usr/bin/clickhouse
15. Poco::PooledThread::run() in /usr/bin/clickhouse
16. Poco::ThreadImpl::runnableEntry(void*) in /usr/bin/clickhouse
17. start_thread in /lib64/libpthread-2.22.so
18. clone in /lib64/libc-2.22.so

how to solve this problem, thx.

**Comment by mo-avatar** [comment]
May be you can enlarge  max_thread_pool_size  to enlarge the thread pool of clilckhouse to supply sufficient threads if your linux still have the resource.

**Reply by shenjingmushan** [reply]
I have enlarged max_thread_pool_size to 40000, but the problem remains.
![image](https://user-images.githubusercontent.com/7822528/217691444-a53510c5-4f40-4026-8171-7eedd066757d.png)

**Reply by mo-avatar** [reply]
select * from xxxx t final order by t.a desc limit 100 settings  max_threads = certain_number. May be you can limit the query with certain number of threads, please lower max_threads  to a low number(default is unlimited) may be it will help.

**Reply by shenjingmushan** [reply]
Thanks, that works by lower max_final_threads to 8.

In addition, i still have the question, when i set max_thread_pool_size=40000, then occur error "Cannot schedule a task: cannot allocate thread", i "select * from system.metrics where metric like '%Global%'", the GlobalThread still 430, not 40000.
Am i misunderstood this parameter, or how to enlarge the GlobalThread?
Thank you for your answers.

**Reply by mo-avatar** [reply]
max_thread_pool_size  is the upper limit , GlobalThread is the actual threads num, so may be it is not limited by max_thread_pool_size  at all, but some other limits in the system.
